### PR TITLE
Creating a script that will generate slack channels based on the team…

### LINKED ISSUE #10: SPIKE - Auto Generate Slack Channels from Framework
Can we auto generate slack channels from the schema? This would prove that organisational communication settings could be controlled from source, meaning it could be deployed to different tooling when or if it ever came about
If so how do we then ensure that the documentation can retrieve those links so we can still maintain that central thin routing mechanism via the links to the right places

### PR DESCRIPTION
**Pull request by AnilKerai**
…config project. 

Resolves #10 

First implementation of a slack channel generator.

### CODE REVIEW

File: app/comms-gen/slack/README.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+This script makes use of the [Slack API](https://api.slack.com/methods/conversations.create). In order to integrate with Slack, we are required to create an app to act as an identity which will make changes to your workspace on your behalf via the API.
+
+## Installing the Team Configurator
+Slack App in review process. 
+
+TODO: This section needs work.
+
+Browse slack apps and install an app called 'Team Configurator'. This acts an identity for the API calls. You will need this apps bot token from the settings, if available. This is potentially accessible in the app settings, but I need to verify this.
+
+## Creating a slack bot app
```
Comment (griff182uk) [inline app/comms-gen/slack/README.md:18]: I noticed there's a bit in slack where you can apply some yaml for the creation of an app (app manifest). Shall we just recommend that as the route so all in code?

e.g. Add below as an app manifest, go get token for this new app. Bang. Done.

```
display_information:
  name: workspace-admin
features:
  bot_user:
    display_name: workspace-admin
    always_online: false
oauth_config:
  scopes:
    bot:
      - channels:manage
      - channels:read
      - groups:read
      - im:read
      - mpim:read
settings:
  org_deploy_enabled: false
  socket_mode_enabled: false
  token_rotation_enabled: false
```

**Review by griff182uk — COMMENTED** [review]
I learned a ton and my ineptitude with APIs has likely proven good from a QA/user testing perspective :D

File: app/comms-gen/slack/README.md (hunk 1 above)
Comment (AnilKerai) [reply app/comms-gen/slack/README.md:18]: Great shout dude, I hadn't spotted you can do this :)

File: app/comms-gen/slack/generate-slack-channels.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
+import os
+from os.path import isfile, join
+import json
+import glob
+import requests
+
+def set_context():
+    currentPath = os.path.dirname(os.path.abspath(__file__))
+    endIndex = currentPath.find('/app')
```
[the remark this answers is not in the document]
Comment (AnilKerai) [reply app/comms-gen/slack/generate-slack-channels.py]: yeah good shout. I'm currently on Linux so i didn't notice. I've updated to account for Windows and i've also aborted the script if that reference folder cannot be found.

File: app/comms-gen/slack/generate-slack-channels.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        print(f'Generating slack channels for {team["name"]}..')
+        for channel in team['channels']:
+            url = baseUrl.format(channel['name'])
+            response = requests.post(url, headers=headers)
+            # response = requests.post('https://slack.com/api/conversations.create?name=python-test', headers=headers)
+            if (not response.ok):
+                print(f'Something went wrong creating {channel["name"]}. Reason: {response.reason}')
+
+            responseBody = response.json()
+            if (not responseBody['ok'] and responseBody['error'].lower() == 'name_taken'):
```
[the remark this answers is not in the document]
Comment (AnilKerai) [reply app/comms-gen/slack/generate-slack-channels.py:119]: done :)

File: app/comms-gen/slack/generate-slack-channels.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    return teams
+
+def generate_slack_channels(apiToken, teamsData):
+    print('Generating slack channels..')
+
+    baseUrl = 'https://slack.com/api/conversations.create?name={}'
+    headers = {'Authorization': f'Bearer {apiToken}'}
+
+    for team in teamsData:
+        print(f'Generating slack channels for {team["name"]}..')
```
[the remark this answers is not in the document]
Comment (AnilKerai) [reply app/comms-gen/slack/generate-slack-channels.py:109]: Done :)

File: app/comms-gen/slack/generate-slack-channels.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+            responseBody = response.json()
+            if (not responseBody['ok'] and responseBody['error'].lower() == 'name_taken'):
+                print(f'{channel["name"]} already exists')
+
+        print(f"Done generating slack channels for {team['name']}.")
+
+    print('Done.')
+
+def main():
```
[the remark this answers is not in the document]
Comment (AnilKerai) [reply app/comms-gen/slack/generate-slack-channels.py:136]: i'll pick this up in a different script bud. It sounds like it's a slightly different set of requirements, albeit with a fair amount of crossover

**Review by griff182uk — APPROVED** [review]
All good now

**Pull request merged**